java/ql/src/experimental/Security/CWE/CWE-470/UnsafeReflectionLib.qll

```diff
@@ -0,0 +1,94 @@
+import java
+import DataFlow
+import semmle.code.java.Reflection
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.DataFlow3
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+
+/**
+ * A call to a Java standard library method which constructs or returns a `Class<T>` from a `String`.
+ * e.g `Class.forName(...)` or `ClassLoader.loadClass(...)`
+ */
+class ReflectiveClassIdentifierMethodAccessCall extends MethodAccess {
+  ReflectiveClassIdentifierMethodAccessCall() {
+    this instanceof ReflectiveClassIdentifierMethodAccess
+  }
+}
+
+/**
+ * Unsafe reflection sink.
+ * e.g `Constructor.newInstance(...)` or `Method.invoke(...)` or `Class.newInstance()`.
+ */
+class UnsafeReflectionSink extends DataFlow::ExprNode {
+  UnsafeReflectionSink() {
+    exists(MethodAccess ma |
+      (
+        ma.getMethod().hasQualifiedName("java.lang.reflect", "Constructor<>", "newInstance")
+        or
+        ma.getMethod().hasQualifiedName("java.lang.reflect", "Method", "invoke")
+      ) and
+      ma.getQualifier() = this.asExpr() and
+      exists(ReflectionArgsConfig rac | rac.hasFlowToExpr(ma.getAnArgument()))
+    )
+  }
+}
+
+/** Taint-tracking configuration tracing flow from remote sources to specifying the initialization parameters to the constructor or method. */
+class ReflectionArgsConfig extends TaintTracking2::Configuration {
+  ReflectionArgsConfig() { this = "ReflectionArgsConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(NewInstance ni | ni.getAnArgument() = sink.asExpr())
+    or
+    exists(MethodAccess ma, ReflectionInvokeObjectConfig rioc |
+      ma.getMethod().hasQualifiedName("java.lang.reflect", "Method", "invoke") and
+      ma.getArgument(1) = sink.asExpr() and
+      rioc.hasFlow(_, DataFlow::exprNode(ma.getArgument(0)))
+    )
+  }
+
+  override predicate isAdditionalTaintStep(Node pred, Node succ) {
+    exists(MethodAccess ma |
+      ma.getMethod().getReturnType().hasName("Object[]") and
```

Comment:
All arguments flow to all Object-array-typed returns? This is very vague. What do you intend this to do?

Reply:
This is an example. The second parameter value type of the `invoke` method is `Object[]`
```java
String parameterValue = request.getParameter("parameterValue");
Object[] methodInput = serializer.deserializeMethodInput(parameterValue, files, method);
Object result = method.invoke(bean, methodInput);
```

Reply:
This is too vague to include. We don't know that parameters to a method taint its return value in this way. Either model particular well-known APIs that you want to propagate taint or drop the model entirely.

Reply:
I don't think so. It is indeed vague for a single configuration, but here I think there will be no false positives.

Reply:
False positive:

```
Object[] taintFromFirst(String toRet, String toLog) {
  Log.log(toLog);
  return new Object[] { toRet };
}

taintFromFirst("not user controlled", getUserControlled());
```

I will not merge this PR with this rule included, or anything like it. We need to have a much stronger reason to think a method propagates taint than an unprincipled guess from its signature.